Added metrics to evaluation

SteffenTunkel opened the pull request:
-add data description metrics: size_training_original, bias_test, bias_training, bias_training_original, prop1_defect, prop1_clean
-add performance evaluation metrics: aucRoI, cost, necm10
-carry before-preprocessing train data through execution plans
-fix typos

SteffenTunkel committed:
Added metrics to evaluation

-add data description metrics: size_training_original, bias_test, bias_training, bias_training_original, prop1_defect, prop1_clean
-add performance evaluation metrics: aucRoI, cost, necm10
-carry before-preprocessing train data through execution plans
-fix typos

sherbold commented on CrossPare/src/main/java/de/ugoe/cs/cpdp/eval/AbstractWekaEvaluation.java:
```diff
@@ -507,4 +538,79 @@ public static double getNECM(Evaluation eval, double factor) {
+     * @return bias
+     */
+    public static double getBias(Instances data) {
+        double count = 0.0;
+        for (Instance instance : data) {
+            if (instance.classValue() > 0.0) {
+                count += 1.0;
+            }
+        }
```
data(data.classIndex()).nominalCounts should already implement this

SteffenTunkel replied:
I changed it to use the nominalCounts attribute. But not all 'Instances' are 'nominal' at that point.

sherbold replied:
Ah, right. In that case your solution was better and I guess you should revert this. Sorry, I forgot about that case. The rest looks good now.

sherbold commented on CrossPare/src/main/java/de/ugoe/cs/cpdp/eval/AbstractWekaEvaluation.java:
```diff
@@ -507,4 +538,79 @@ public static double getNECM(Evaluation eval, double factor) {
+            }
+        }
+        effortDefect.sort(Collections.reverseOrder());
+        double percentile = Math.ceil(effortDefect.size() / 100.0);
```
should use int as datatype, since you are actually calculating an index here. maybe also rename to percentile_index?

sherbold commented on CrossPare/src/main/java/de/ugoe/cs/cpdp/eval/AbstractWekaEvaluation.java:
```diff
@@ -507,4 +538,79 @@ public static double getNECM(Evaluation eval, double factor) {
+        }
+        effortDefect.sort(Collections.reverseOrder());
+        double percentile = Math.ceil(effortDefect.size() / 100.0);
+        double LargestPercentileEffort = 0.0;
```
NIT: lowercase first letter

(NIT for nitpick, ie. usually minor style issues)

sherbold commented on CrossPare/src/main/java/de/ugoe/cs/cpdp/eval/AbstractWekaEvaluation.java:
```diff
@@ -507,4 +538,79 @@ public static double getNECM(Evaluation eval, double factor) {
+            }
+        }
+        effortClean.sort(Collections.reverseOrder());
+        double percentile = Math.ceil(effortClean.size() / 100.0);
+        double LargestPercentileEffort = 0.0;
```
see above

sherbold commented on CrossPare/src/main/java/de/ugoe/cs/cpdp/eval/EffortMetricCalculator.java:
```diff
@@ -123,6 +124,59 @@ public double getAUCEC() {
+            numPositives += 1;
+            cumLabel[0] = 1;
+        }
+        else cumLabel[0] = 0;
```
avoid if/else without {}

sherbold commented on CrossPare/src/main/java/de/ugoe/cs/cpdp/execution/AbstractCrossProjectExperiment.java:
```diff
@@ -181,6 +181,7 @@ public void run() {
                     versionCount++;
                     continue;
                 }
+                SetUniqueList<SoftwareVersion> trainversionOriginalSet = SetUniqueList.setUniqueList(trainversionSet);
```
are you sure this is a deep copy and the subsequent changes do not affect this copy? I have not checked myself, so I really do not know. please just double check because this would be annoying. Easiest way would be to just use Undersampling or Oversampling and see if the values are as expected.

sherbold replied:
a different solution may be just to pull up the "makeSingleVersionSet" for the copy and do this here instead of in line 225, because that should be a deep copy if I remember correctly.

SteffenTunkel replied:
Yes, I checked it with oversampling beforehand. Anyway, your idea seems to be even cleaner.

sherbold commented on CrossPare/src/main/java/de/ugoe/cs/cpdp/execution/WithinProjectOrderedSplitExperiment.java:
```diff
@@ -164,7 +164,7 @@ public void run() {
                     }
                 }
-
+                SoftwareVersion trainversionOriginal = new SoftwareVersion(testversion);
```
is testversion here correct? the code above splits the instances based on their index, the training data should be in the trainversion.

sherbold reviewed (commented):
Looks mostly good, but I have some smaller comments. 

Note: I did not check the math for the aucROI

SteffenTunkel committed:
Fix issues in pull request

SteffenTunkel committed:
Change getBias function

SteffenTunkel committed:
add SMOTUNED Filter and OutOfSampleBootstrap as ExecStrat and Selector

sherbold commented on CrossPare/src/main/java/de/ugoe/cs/cpdp/dataselection/OutOfSampleBootstrap.java:
```diff
@@ -0,0 +1,152 @@
+     */
+    @Override
+    public SoftwareVersion apply(SoftwareVersion testversion, SoftwareVersion trainversion) {
+        WekaUtils.makeClassBinary(testversion.getInstances());
+        WekaUtils.makeClassBinary(trainversion.getInstances());
```
avoid side effects. rather let this crash with an exception if the class is not yet binary.

sherbold commented on CrossPare/src/main/java/de/ugoe/cs/cpdp/dataselection/OutOfSampleBootstrap.java:
```diff
@@ -0,0 +1,152 @@
+                abortCounter += 1;
+                if (abortCounter >= abortLimit){
+                    throw new RuntimeException(
+                            String.format("Error while taking a bootstrap sample of %s: After %d iterations, " +
+                                            "no bootstrap sample containing a defect was created.",
```
this message is not accurage. there may be defects, but not more than minDefects

sherbold commented on CrossPare/src/main/java/de/ugoe/cs/cpdp/dataselection/OutOfSampleBootstrap.java:
```diff
@@ -0,0 +1,152 @@
+                continue;
+            }
+
+            List<Double> leftovers = new ArrayList<>();
```
this should be `List<Integer>`, then you can also delete some casting below

sherbold commented on CrossPare/src/main/java/de/ugoe/cs/cpdp/execution/OutOfSampleBootstrapExperiment.java:
```diff
@@ -0,0 +1,414 @@
+// Copyright 2015 Georg-August-Universität Göttingen, Germany
```
I like the filter better, but we need to discuss memory consumption

sherbold commented on CrossPare/src/main/java/de/ugoe/cs/cpdp/dataprocessing/SMOTUNED.java:
```diff
@@ -0,0 +1,638 @@
+    /**
+     * maximal ratio of the final amount of majority class instances to the overall instances (1.0 = no undersampling)
+     */
+    final double undersampleRate = 0.5;
```
empty line missing

sherbold commented on CrossPare/src/main/java/de/ugoe/cs/cpdp/dataprocessing/SMOTUNED.java:
```diff
@@ -0,0 +1,638 @@
+            if (this.numNeighbors >= sizeMinority){
+                this.numNeighbors = sizeMinority - 1;
+            }
+            ArrayList<Integer> synthList = new ArrayList<Integer>();
```
general style hint: 
avoid `ArrayList` and `LinkedList` as types, always use `List` instead. this makes the code easier to maintain and switch types (same for sets, etc).

sherbold reviewed (commented):
looks pretty good already

SteffenTunkel committed:
fix according to PR feedback

SteffenTunkel committed:
change orginal training data bypass position

For the experiment the original training data needs to be taken at least after the setwise selector (TestAsTraining) to contain reasonable values.

SteffenTunkel committed:
add AUC Alberg metric

SteffenTunkel committed:
change training data orginal bypass dependent on TestAsTraining

sherbold reviewed (commented):
LGTM 👍
